tendermint/src/error.rs

```diff
@@ -1,206 +1,190 @@
-
-    /// Invalid Evidence
-    #[error("invalid evidence")]
-    InvalidEvidence,
-
-    /// Invalid BlockIdFlag
-    #[error("invalid block id flag")]
-    BlockIdFlag,
-
-    /// Negative voting power
-    #[error("negative power")]
-    NegativePower,
-
-    /// Mismatch between raw voting power and computed one in validator set
-    #[error("mismatch between raw voting power ({raw}) and computed one ({computed})")]
-    RawVotingPowerMismatch {
-        /// raw voting power
-        raw: vote::Power,
-        /// computed voting power
-        computed: vote::Power,
-    },
-
-    /// Missing Public Key
-    #[error("missing public key")]
-    MissingPublicKey,
-
-    /// Invalid validator parameters
-    #[error("invalid validator parameters")]
-    InvalidValidatorParams,
-
-    /// Invalid version parameters
-    #[error("invalid version parameters")]
-    InvalidVersionParams,
-
-    /// Negative max_age_num_blocks in Evidence parameters
-    #[error("negative max_age_num_blocks")]
-    NegativeMaxAgeNum,
-
-    /// Missing max_age_duration in evidence parameters
-    #[error("missing max_age_duration")]
-    MissingMaxAgeDuration,
-
-    /// Proposer not found in validator set
-    #[error("proposer with address '{}' not found in validator set", _0)]
-    ProposerNotFound(account::Id),
-}
+        InvalidHashSize
+        |_| { format_args!("invalid hash: expected hash size to be 32 bytes") },
+
+        NoTimestamp
+        |_| { format_args!("no timestamp") },
+
+        InvalidTimestamp
+        [DisplayError<Error>]
+        |_| { format_args!("invalid timestamp") },
+
+        InvalidAccountIdLength
+        |_| { format_args!("invalid account ID length") },
+
+        InvalidSignatureIdLength
+        |_| { format_args!("invalid signature ID length") },
+
+        IntegerOverflow
+        |_| { format_args!("integer overflow") },
+
+        NoVoteFound
+        |_| { format_args!("no vote found") },
+
+        NoProposalFound
+        |_| { format_args!("no proposal found") },
+
+        InvalidAppHashLength
+        |_| { format_args!("invalid app hash length") },
+
+        InvalidPartSetHeader
+        [DisplayError<Error>]
+        |_| { format_args!("invalid part set header") },
+
+        MissingHeader
+        |_| { format_args!("missing header field") },
+
+        MissingData
+        |_| { format_args!("missing data field") },
+
+        MissingEvidence
+        |_| { format_args!("missing evidence field") },
+
+        MissingTimestamp
+        |_| { format_args!("missing timestamp field") },
+
+        InvalidBlock
+        [DisplayError<Error>]
+        |_| { format_args!("invalid block") },
+
+        InvalidFristBlock
+        [DisplayError<Error>]
+        |_| { format_args!("invalid first block") },
+
+        MissingVersion
+        |_| { format_args!("missing version") },
+
+        InvalidHeader
+        [DisplayError<Error>]
+        |_| { format_args!("invalid header") },
+
+        InvalidFirstHeader
```

Comment:
This error do not have a source and is used in one place. So it can be

```rust
InvalidFirstHeader
  |_| { format_args!("last_block_id is not null on first height") }
```